### PR TITLE
[REVIEW] Slurm containerized setup

### PR DESCRIPTION
**Pull request by beckernick**
This PR:
- Includes a variety of configuration and setup for running in containers with Slurm, in this case oriented toward an internal NVIDIA cluster
- Makes some small tweaks to the load_test script
- Adds a `wait.py`, which can be used in a multi-node setup to ensure the client process on the scheduler node doesn't begin until all dask workers have attached to the cluster
- Updates the default UNIX permissions of a large number of files

### CODE REVIEW

File: tpcx_bb/benchmark_runner.py [hunk 1]
```diff
@@ -13,9 +13,61 @@ def get_qnum_from_filename(name):
...
+        #"15",
+        #"16",
+        #"17",
+        #"18",
+        #"19",
+        #"20",
+        #"21",
+        #]
+
+dask_qnums = ["01", "08"]
```
Comment (ayushdg) [inline tpcx_bb/benchmark_runner.py:32]: Is this intentional? I also see 08, 01 for blazing below

File: tpcx_bb/benchmark_runner/draco/README.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+The Docker images used are date tagged but not intended to signify "stable" images.
+
+Setup:
+1. Somewhere you have Docker permissions, build image:
+```bash
+cd $TPCX_BB/tpcx_bb/benchmark_runner/draco
+docker build -t draco .
+# update image id
+docker tag bb38976d03cf dockerhub_username/draco:8_24_0
```
Comment (ayushdg) [inline tpcx_bb/benchmark_runner/draco/README.md:11]: `bb38976d03cf` is this a random image_id? Maybe replace with {image_tag} or something similar?

File: tpcx_bb/benchmark_runner/draco/run.sh [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+IMAGE=nickb500/draco-tpcxbb:2020_09_10
+
+DATA_PATH="/fs/sjc1-gcl01/datasets/tpcx-bb"
+TPCX_BB_HOME="/root/tpcx-bb"
+
+
+rm *.out
+rm -rf $HOME/dask-local-directory/*
+
+nvs batch \
```
Comment (ayushdg) [inline tpcx_bb/benchmark_runner/draco/run.sh:17]: Does `sbatch` work? Would be a more generic slurm approach with it

File: tpcx_bb/benchmark_runner/selene/README.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+Setup:
+1. Somewhere you have Docker permissions, build image:
+```bash
+cd $TPCX_BB/tpcx_bb/benchmark_runner/draco
+docker build -t draco .
+# update image id
+docker tag bb38976d03cf dockerhub_username/draco:8_24_0
+docker push dockerhub_username/draco
+```
```
Comment (ayushdg) [inline tpcx_bb/benchmark_runner/selene/README.md:13]: Update to point to `selene`

File: tpcx_bb/benchmark_runner/selene/README.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+2. In Draco, create a dask-local-directory. This is a network share, accessible to all nodes:
+```bash
+mkdir -p ~/dask-local-directory
+```
+
+3. Update bash vars `run.sh` to make sure you're specifying the right image and number of workers
+4. Run:
+```bash
+bash benchmark_runner/draco/run.sh
```
Comment (ayushdg) [inline tpcx_bb/benchmark_runner/selene/README.md:23]: same as above

File: tpcx_bb/benchmark_runner/selene/rapids-tpcx-bb-cuda11.yml [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  - psutil
+  - ipykernel
+  - jupyterlab
+  - gspread
+  - oauth2client
+  - pip
+  - pip:
+    - jupyter-server-proxy
+    - git+https://github.com/dask/dask.git
+    - git+https://github.com/dask/distributed.git
```
Comment (ayushdg) [inline tpcx_bb/benchmark_runner/selene/rapids-tpcx-bb-cuda11.yml:33]: May not be relevant in general but may want to pin these to a specific commit if needed

File: tpcx_bb/cluster_configuration/cluster-startup-selene.sh [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+USERNAME=$(whoami)
+# HOSTNAME=$(hostname -i)
+HOSTNAME=$HOSTNAME
+
+MAX_SYSTEM_MEMORY=$(free -m | awk '/^Mem:/{print $2}')M
+DEVICE_MEMORY_LIMIT="32GB"
+POOL_SIZE="36GB"
+
+TPCX_BB_HOME=$HOME/tpcx-bb
+CONDA_ENV_NAME="rapids"
```
Comment (ayushdg) [inline tpcx_bb/cluster_configuration/cluster-startup-selene.sh:16]: Do you think it makes sense to generalize the env name across draco, selene, cluster_startup.sh to a common name rapids-tpcx-bb

**Review by ayushdg — CHANGES_REQUESTED** [review]
Minor feedback comments

File: tpcx_bb/benchmark_runner/selene/README.md (hunk 4 above)
Comment (beckernick) [reply tpcx_bb/benchmark_runner/selene/README.md:13]: Going to remove this.

File: tpcx_bb/benchmark_runner/selene/rapids-tpcx-bb-cuda11.yml (hunk 6 above)
Comment (beckernick) [reply tpcx_bb/benchmark_runner/selene/rapids-tpcx-bb-cuda11.yml:33]: For now, going to leave unpinned as we're keeping a history of the containers.

File: tpcx_bb/cluster_configuration/cluster-startup-selene.sh (hunk 7 above)
Comment (beckernick) [reply tpcx_bb/cluster_configuration/cluster-startup-selene.sh:16]: I think so. Will standardize on rapids-tpcx-bb

File: tpcx_bb/xbb_tools/cluster_startup.py [hunk 8]
```diff
@@ -66,6 +66,15 @@ def attach_to_cluster(config, create_blazing_context=False):
     scheduler_file = config.get("scheduler_file_path")
     host = config.get("cluster_host")
     port = config.get("cluster_port", "8786")
+    
+    if scheduler_file is not None:
```
Comment (beckernick) [inline tpcx_bb/xbb_tools/cluster_startup.py:70]: To be removed. Missing while resolving merge conflicts

**Comment by beckernick** [comment]
This should be ready for review @ayushdg @VibhuJawa

**Review by ayushdg — APPROVED** [review]
LGTM, thank you so much for these updates @beckernick. Should make it much easier to setup.

**Comment by beckernick** [comment]
Made one small update to switch the load test chunk size back to 128 MiB from 1 GB, to ease running on more constrained systems.

**Pull request merged**